nifi-nar-bundles/nifi-salesforce-bundle/nifi-salesforce-processors/src/main/java/org/apache/nifi/processors/salesforce/QuerySalesforceObject.java

```diff
@@ -0,0 +1,505 @@
+            results.add(
+                    new ValidationResult.Builder()
+                            .subject(AGE_FIELD.getDisplayName())
+                            .valid(false)
+                            .explanation("it requires " + TIMESTAMP_FORMAT.getDisplayName() + " also to be set.")
+                            .build()
+            );
+        }
+
+        if (validationContext.getProperty(INITIAL_AGE_FILTER).isSet() && !validationContext.getProperty(AGE_FIELD).isSet()) {
+            results.add(
+                    new ValidationResult.Builder()
+                            .subject(INITIAL_AGE_FILTER.getDisplayName())
+                            .valid(false)
+                            .explanation("it requires " + AGE_FIELD.getDisplayName() + " also to be set.")
+                            .build()
+            );
+        }
+
+        return results;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        String sObject = context.getProperty(SOBJECT_NAME).getValue();
+        String fields = context.getProperty(FIELD_NAMES).getValue();
+        String customWhereClause = context.getProperty(CUSTOM_WHERE_CONDITION).getValue();
+        String dateFormat = context.getProperty(DateTimeUtils.DATE_FORMAT).getValue();
+        String timeFormat = context.getProperty(DateTimeUtils.TIME_FORMAT).getValue();
+        String timestampFormat = context.getProperty(TIMESTAMP_FORMAT).getValue();
+        RecordSetWriterFactory writerFactory = context.getProperty(RECORD_WRITER).asControllerService(RecordSetWriterFactory.class);
+        boolean includeZeroRecordFlowFiles = context.getProperty(INCLUDE_ZERO_RECORD_FLOWFILES).asBoolean();
+
+        String ageField = context.getProperty(AGE_FIELD).getValue();
+        String initialAgeFilter = context.getProperty(INITIAL_AGE_FILTER).getValue();
+        Long ageDelayMs = context.getProperty(AGE_DELAY).asTimePeriod(TimeUnit.MILLISECONDS);
+
+        String ageFilterLower;
+        StateMap state;
+        try {
+            state = context.getStateManager().getState(Scope.CLUSTER);
+            ageFilterLower = state.get(LAST_AGE_FILTER);
+        } catch (IOException e) {
+            throw new ProcessException("Last Age Filter state retrieval failed", e);
+        }
+
+        String ageFilterUpper;
+        if (ageField == null) {
+            ageFilterUpper = null;
+        } else {
+            Instant ageFilterUpperTime;
+            if (ageDelayMs == null) {
+                ageFilterUpperTime = Instant.now();
+            } else {
+                ageFilterUpperTime = Instant.now().minus(ageDelayMs, ChronoUnit.MILLIS);
+            }
+            SimpleDateFormat simpleDateFormat = new SimpleDateFormat(timestampFormat);
+            simpleDateFormat.setTimeZone(TimeZone.getTimeZone("UTC"));
+            ageFilterUpper = simpleDateFormat.format(Date.from(ageFilterUpperTime));
```

Comment:
@Lehel44, can you provide some additional background on the change from DateTimeFormatter to the older SimpleDateFormat?  The DateTimeFormatter should be a better for parsing to an Instant.

In addition, hard-coding the `UTC` timezone seems problematic.  The timestamp format can include a timezone element, so it seems like it would better to avoid hard-coding a particular timezone.  There may be other reasons for this, but at minimum this should be documented in the timestamp format descriptor.

Reply:
I think you can't see my outdated reply.

"Using DateTimeFormatter with the LocalDate pattern can result in the full form of localized offset concatenated to the formatted date string which salesforce cannot handle.

The timestampFromat property is derived from the DateTimeUtils.timeStampFormat property which specifies that the value must match the Java Simple Date Format. Since DateTimeFormatter uses different format patterns, the following default pattern "yyyy-MM-dd'T'HH:mm:ss.SSSZZZZ" would result in two different date strings.

With LocalDateTime: 2022-03-11T00:34:40.210+0100
With DateTimeFormatter: 2022-03-11T00:34:40.210GMT+01:00
From the DateTimeFormatter documentation: Offset Z: "Four letters outputs the full form of localized offset".

I can change the java.sql.TimeStamp.from to Date.from."

Reply:
About time zones:

1. Salesforce uses default UTC standard for timezones. This change will recalculate any time in a timezone according to the UTC standard time, e.g. time in EDT -> time in UTC-4.
2. There is no user-provided timestamp in among the processor properties. The timestamp formatter only applies for the instant created in the code and is configurable in case it can also be configured in Salesforce.

Reply:
So in a nutshell
-  The timestamp formatter processor property is derived from a property which was meant to use LocalDateTime format (has built-in validator).
- We create a datetime string for the SOQL query which Salesforce accepts, so we use LocalDateTime to omit the Zone Id.
- We might create a different timestamp formatter property with SimpleDateFormat pattern, or we use LocalDateTime in this case which is I think easier.
- We convert the date time to UTC format, to make sure Salesforce can handle it